ecosystem/indexer-grpc/indexer-grpc-cache-worker/src/worker.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    constants::{
+        APTOS_INDEXER_CACHE_RECONNECT_LIMIT, APTOS_INDEXER_FULLNODE_GRPC_PORT,
+        APTOS_INDEXER_GRPC_CACHE_PORT, APTOS_INDEXER_GRPC_CACHE_WORKER_CONFIG_PATH_VAR,
+        CACHE_KEY_CHAIN_ID,
+    },
+    get_ttl_in_seconds, IndexerGrpcCacheWorkerConfig,
+};
+use aptos_logger::{error, info};
+use aptos_protos::datastream::v1 as datastream;
+use deadpool_redis::{redis::cmd, Config, Pool, Runtime};
+use futures::{self, StreamExt};
+use moving_average::MovingAverage;
+use std::{path::PathBuf, sync::Arc};
+
+pub fn get_worker_config_file_path() -> String {
+    std::env::var(APTOS_INDEXER_GRPC_CACHE_WORKER_CONFIG_PATH_VAR)
+        .expect("WORKER_CONFIG_PATH is required.")
+}
+
+pub struct Worker {
+    redis_pool: Arc<Pool>,
+    config: IndexerGrpcCacheWorkerConfig,
+    next_version: Option<u64>,
+    retry_count: u64,
+}
+
+impl Worker {
+    pub async fn new() -> Self {
+        let config_path = get_worker_config_file_path();
+        let config = IndexerGrpcCacheWorkerConfig::load(PathBuf::from(config_path)).unwrap();
+        let redis_address = match &config.redis_address {
+            Some(addr) => addr.clone(),
+            _ => "127.0.0.1".to_string(),
+        };
+
+        let cfg = Config::from_url(format!(
+            "redis://{}:{}",
+            redis_address, APTOS_INDEXER_GRPC_CACHE_PORT
+        ));
+
+        let redis_pool = Arc::new(cfg.create_pool(Some(Runtime::Tokio1)).unwrap());
+
+        let retry_count = 0;
+
+        let next_version = config.starting_version.as_ref().copied();
+
+        Self {
+            redis_pool,
+            config,
+            next_version,
+            retry_count,
+        }
+    }
+
+    /// Sync with the cache to verify and process the next version.
+    async fn cache_status_sync(&mut self) {
+        // Failure to connect to Redis is a fatal error.
+        let mut conn = self.redis_pool.get().await.unwrap();
+
+        // 1. If the chain id is not present in the cache, set it.
+        let if_chain_id_set: bool = cmd("SETNX")
+            .arg(&[
+                CACHE_KEY_CHAIN_ID.to_string(),
+                self.config.chain_id.to_string(),
+            ])
+            .query_async(&mut conn)
+            .await
+            .expect("[Redis] SETNX the chain id.");
+
+        // 2. If the chain id is present, check if it matches the config.
+        if !if_chain_id_set {
+            let cache_chain_id: String = cmd("GET")
+                .arg(&[CACHE_KEY_CHAIN_ID.to_string()])
+                .query_async(&mut conn)
+                .await
+                .expect("[Redis] Set the chain id.");
+            assert_eq!(cache_chain_id, self.config.chain_id.to_string());
+        }
```

Comment:
let's read from the filestore for chain id as well

Reply:
Going to add in the follow-up PR.